packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts

```diff
@@ -0,0 +1,251 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as cxschema from '@aws-cdk/cloud-assembly-schema';
+import { FileAssetSource, FileAssetPackaging, DockerImageAssetSource } from '../assets';
+import { Stack } from '../stack';
+import { resolvedOr } from './_shared';
+import { ISynthesisSession } from './types';
+
+/**
+ * Build an asset manifest from assets added to a stack
+ *
+ * This class does not need to be used by app builders; it is only nessary for building Stack Synthesizers.
+ */
+export class AssetManifestBuilder {
+  private readonly files: NonNullable<cxschema.AssetManifest['files']> = {};
+  private readonly dockerImages: NonNullable<cxschema.AssetManifest['dockerImages']> = {};
+
+  /**
+   * Add a file asset to the manifest with default settings
+   *
+   * Derive the region from the stack, use the asset hash as the key, copy the
+   * file extension over, and set the prefix.
+   */
+  public defaultAddFileAsset(stack: Stack, asset: FileAssetSource, target: AssetManifestFileDestination) {
+    validateFileAssetSource(asset);
+
+    const extension =
+      asset.fileName != undefined ? path.extname(asset.fileName) : '';
+    const objectKey =
+      (target.bucketPrefix ?? '') +
+      asset.sourceHash +
+      (asset.packaging === FileAssetPackaging.ZIP_DIRECTORY
+        ? '.zip'
+        : extension);
+
+    // Add to manifest
+    return this.addFileAsset(stack, asset.sourceHash, {
+      path: asset.fileName,
+      executable: asset.executable,
+      packaging: asset.packaging,
+    }, {
+      bucketName: target.bucketName,
+      objectKey,
+      region: resolvedOr(stack.region, undefined),
+      assumeRoleArn: target.role?.assumeRoleArn,
+      assumeRoleExternalId: target.role?.assumeRoleExternalId,
+    });
+  }
+
+  /**
+   * Add a docker image asset to the manifest with default settings
+   *
+   * Derive the region from the stack, use the asset hash as the key, and set the prefix.
+   */
+  public defaultAddDockerImageAsset(
+    stack: Stack,
+    asset: DockerImageAssetSource,
+    target: AssetManifestDockerImageDestination,
+  ) {
+    validateDockerImageAssetSource(asset);
+    const imageTag = `${target.dockerTagPrefix ?? ''}${asset.sourceHash}`;
+
+    // Add to manifest
+    return this.addDockerImageAsset(stack, asset.sourceHash, {
+      executable: asset.executable,
+      directory: asset.directoryName,
+      dockerBuildArgs: asset.dockerBuildArgs,
+      dockerBuildTarget: asset.dockerBuildTarget,
+      dockerFile: asset.dockerFile,
+      networkMode: asset.networkMode,
+      platform: asset.platform,
+    }, {
+      repositoryName: target.repositoryName,
+      imageTag,
```

Comment:
[Q] This is equivalent to `imageTag: imageTag,`?

Reply:
Yes!